src/SaveRelationsBehavior.php

```diff
@@ -298,29 +299,40 @@ public function afterSave()
                                 if ($relation->via !== null) {
                                     $relationModel->save(false);
                                 }
-                                $model->link($relationName, $relationModel);
+                                $junctionTableColumns = $this->_getJunctionTableColumns($relationName, $relationModel);
+                                $model->link($relationName, $relationModel, $junctionTableColumns);
                             } else {
                                 $existingRecords[] = $relationModel;
                             }
                             if (count($relationModel->dirtyAttributes)) {
                                 $relationModel->save(false);
                             }
                         }
+                        $junctionTablePropertiesUsed = array_key_exists($relationName, $this->junctionTableColumns);
                         // Process existing added and deleted relations
-                        list($addedPks, $deletedPks) = $this->_computePkDiff($this->_oldRelationValue[$relationName], $existingRecords);
+                        list($addedPks, $deletedPks) = $this->_computePkDiff(
+                            $this->_oldRelationValue[$relationName],
+                            $existingRecords,
+                            $junctionTablePropertiesUsed
+                        );
                         // Deleted relations
                         $initialModels = ArrayHelper::index($this->_oldRelationValue[$relationName], function (BaseActiveRecord $model) {
                             return implode("-", $model->getPrimaryKey(true));
                         });
+                        $initialRelations = $model->{$relationName};
                         foreach ($deletedPks as $key) {
                             $model->unlink($relationName, $initialModels[$key], true);
```

Comment:
btw this `unlink()` is a bit redundant. it does not take `$junctionTableColumns` into account, so it can delete all the relations to a single entry.
see https://github.com/yiisoft/yii2/issues/14598

Reply:
@1allen What do you mean by "redundant"?

Reply:
@juban I mean that it could unlink more records than actually expected. You can see an example in the yii2 issue I've mentioned above. Sorry, I've no idea how to fix it at this moment

Reply:
@1allen Indeed, the behavior cannot handle that kind of complex scenario. And as long as the unlink method doesn't handle it either, I don't see an easy way to solve this without jeopardizing the behavior compatibility for different kind of DBMS.